modules/videopress/class.videopress-gutenberg.php

```diff
@@ -0,0 +1,84 @@
+<?php
+
+class VideoPress_Gutenberg {
+
+	/**
+	 * Initialize the VideoPress Gutenberg extension
+	 */
+	public static function init() {
+		if ( self::should_initialize() ) {
+			add_action( 'init', array( __CLASS__, 'register_video_block_with_videopress' ) );
+		}
+	}
+
+	/**
+	 * Check whether conditions indicate the VideoPress Gutenberg extension should be initialized
+	 *
+	 * @return bool
+	 */
+	public static function should_initialize() {
+		// Should not initialize if Gutenberg is not available
+		if ( ! function_exists( 'register_block_type' ) ) {
+			return false;
+		}
+
+		// Should initialize if this is a WP.com site
+		if ( defined( 'IS_WPCOM' ) && IS_WPCOM ) {
+			return true;
+		}
+
+		// Should not initialize if this is a Jetpack site but Jetpack is not active (unless the dev mode is enabled)
+		if ( ! Jetpack::is_active() && ! Jetpack::is_development_mode() ) {
+			return false;
+		}
+
+		// Should initialize by default
+		return true;
+	}
+
+	/**
+	 * Register the Jetpack Gutenberg extension that adds VideoPress support to the core video block.
+	 */
+	public static function register_video_block_with_videopress() {
+		register_block_type( 'core/video', array(
+			'render_callback' => array( __CLASS__, 'render_video_block_with_videopress' ),
+		) );
+	}
+
+	/**
+	 * Render the core video block replacing the src attribute with the VideoPress URL
+	 *
+	 * @param array  $attributes Array containing the video block attributes.
+	 * @param string $content    String containing the video block content.
+	 *
+	 * @return string
+	 */
+	public static function render_video_block_with_videopress( $attributes, $content ) {
+		if ( ! isset( $attributes['id'] ) ) {
+			return $content;
+		}
+
+		$blog_id = get_current_blog_id();
+		$post_id = absint( $attributes['id'] );
+		$videopress_id = video_get_info_by_blogpostid( $blog_id, $post_id )->guid;
+		$videopress_data = videopress_get_video_details( $videopress_id );
+
+		if ( empty( $videopress_data->files->hd->mp4 ) ) {
+			return $content;
+		}
+
+		$videopress_url = $videopress_data->file_url_base->https . $videopress_data->files->hd->mp4;
+
+		return preg_replace(
+			'/src="([^"]+)/',
+			sprintf(
+				'src="%1$s',
+				esc_attr( $videopress_url )
+			),
+			$content,
+			1
+		);
+	}
```

Comment:
While here I have a point I thought we might address, which is to add the original movie source. I noticed that we do that with images that get replaced by WordPress and it seems like a valid thing to do here. On one hand I don't want to delay this work but on the other hand it seems like it would be a convenient place to make the change.

We'd have to _either_ update the RegEx pattern or switch to real HTML parsing, which honestly would have other benefits but obviously would delay the work.

```php
preg_replace(
	'/(\s)(?:src=(['"])(?:(?!\2.)+?)\2(\s)/',
	sprintf(
		'\1src="%1$s" data-orig-file="%2$s"\3',
		esc_url_raw( $videopress_url ),
		esc_url_raw( $original_url ),
	)
);
```

Reply:
Sounds like a good enhancement! However I'm having hard times try to get the original url from the regular expression. I think we should get it from a matched group, but it seems that a group starting with `?:` is a non capturing group, so I don't know how can we define a 3rd group that captures the original src attribute value.

Reply:
> a group starting with ?: is a non capturing group

that's correct. we can make it a capture group by removing the `?:` which tells it to ignore the group.

`(\s)(?:src=(['"]))((?:(?!\2).)+?)\2([\s\/>])`

note that I think I made a mistake in the original one there - [see it live at Regex101](https://regex101.com/r/CMSI64/1)

once we have the original too we could always determine if it needs replacement and early-abort again. if we do this before requesting the metadata we can abort before running those more expensive queries. this would address the case where the media inserter adds the VideoPress URLs already…

```php
function replace_src_preg_callback( $matches ) {
	list( $full_match, $prefix_space, $quote_char, $src, $suffix ) = $matches;

	if ( 0 === strstr( $src, 'https://videos.files.wordpress.com/' ) ) {
		return $full_match;
	}

	$blog_id = …
	$post_id = …

	$videopress_data = …

	if ( empty( $videopress_data->… ) ) {
		return $full_match;
	}

	return $prefix_space . 'src="' . $new_src . '"' . $suffix;
}

…

preg_replace_callback(
	self::VIDEO_SOURCE_PATTERN,
	array( 'self', 'replace_src_preg_callback' ),
	$content,
	/* limit */ 1
); 
```

Reply:
noting how complicated this RegExp pattern is getting even while matching a very simple thing is why I like to encourage us to explore methods that are reliable and more tuned to actual HTML parsing, even though they might sound scary - with small input documents I suspect they bring no performance issues, plug we can preserve the existing markup and not worry about oddball cases such as when we encounter `<video bad-src="bad.mp4" src="good.mp4" />`

Reply:
Ah, I asked a similar question in D23482-code, and it looks like this is quickly increasing in complexity. 

Provided that the related HTML parser isn't prohibitively slow, I think it's safer to use one rather than carry on with the regex matching. Within this function we have a reasonable guarantee that we only are passed the block html?

Reply:
we have a full guarantee to only get the block contents. note too that we could also filter the block were we to want to, but `render_callback` is also sufficient. the only time we'd expect to get lots of HTML is if someone manually makes the caption enormous.

Reply:
not sure cc: @dereksmart @sirreal - most of what I've seen is RegExp work but also most of what I've seen is at risk for these same kind of risks of complex cases defeating the RegExp _or_ of making the RegExp pattern hard to read in an attempt at catching those complex cases

Reply:
Since this is dragging out a bit, let's go back to this as a follow up. By not embedding the original src url, is this something that's lost forever or can we look that up through alternative means?

Reply:
we can avoid the concerns for now and stick to basic replacement

Reply:
> By not embedding the original src url, is this something that's lost forever or can we look that up through alternative means?

It'll be still on the block content stored on the post content.

That's why I think we should not worry about the concerns because we're going to change the post content anyway in future iterations when the block content will render the VideoPress player instead of the native `video` HTML5 player.